src/volatile_memory.rs

```diff
@@ -0,0 +1,656 @@
+        VolatileSlice {
+            addr: null_mut(),
+            size: 0,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a> VolatileSlice<'a> {
+    /// Creates a slice of raw memory that must support volatile access.
+    ///
+    /// To use this safely, the caller must guarantee that the memory at `addr` is `size` bytes long
+    /// and is available for the duration of the lifetime of the new `VolatileSlice`. The caller
+    /// must also guarantee that all other users of the given chunk of memory are using volatile
+    /// accesses.
+    pub unsafe fn new(addr: *mut u8, size: usize) -> VolatileSlice<'a> {
+        VolatileSlice {
+            addr,
+            size,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Gets the address of this slice's memory.
+    pub fn as_ptr(&self) -> *mut u8 {
+        self.addr
+    }
+
+    /// Gets the size of this slice.
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    /// Creates a copy of this slice with the address increased by `count` bytes, and the size
+    /// reduced by `count` bytes.
+    pub fn offset(self, count: usize) -> Result<VolatileSlice<'a>> {
+        let new_addr =
+            (self.addr as usize)
+                .checked_add(count)
+                .ok_or(VolatileMemoryError::Overflow {
+                    base: self.addr as usize,
+                    offset: count,
+                })?;
+        if new_addr > usize::MAX {
+            return Err(VolatileMemoryError::Overflow {
+                base: self.addr as usize,
+                offset: count,
+            })?;
+        }
+        let new_size = self
+            .size
+            .checked_sub(count)
+            .ok_or(VolatileMemoryError::OutOfBounds { addr: new_addr })?;
+        // Safe because the memory has the same lifetime and points to a subset of the memory of the
+        // original slice.
+        unsafe { Ok(VolatileSlice::new(new_addr as *mut u8, new_size)) }
+    }
+
+    /// Copies `self.size()` or `buf.len()` times the size of `T` bytes, whichever is smaller, to
+    /// `buf`.
+    ///
+    /// The copy happens from smallest to largest address in `T` sized chunks using volatile reads.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::fs::File;
+    /// # use std::path::Path;
+    /// # use memory_model::VolatileMemory;
+    /// # fn test_write_null() -> Result<(), ()> {
+    /// let mut mem = [0u8; 32];
+    /// let mem_ref = &mut mem[..];
+    /// let vslice = mem_ref.get_slice(0, 32).map_err(|_| ())?;
+    /// let mut buf = [5u8; 16];
+    /// vslice.copy_to(&mut buf[..]);
+    /// for v in &buf[..] {
+    ///     assert_eq!(buf[0], 0);
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn copy_to<T>(&self, buf: &mut [T])
+    where
+        T: DataInit,
+    {
+        let mut addr = self.addr;
+        for v in buf.iter_mut().take(self.size / size_of::<T>()) {
+            unsafe {
```

Comment:
@zachreizner This code comes from crosvm, can you reply to @alexandruag?

I agree that we might want to use compiler barrier + read_unaligned instead of read_volatile. However I think we should keep it as is and fix it later, to avoid delaying this pull request further.

Reply:
Fixed now via `#[repr(packed)]`.